### PR TITLE
Implement a method that extracts all Python package dependencies

### PR DESCRIPTION
**Pull request by fuhuifang**
- [x] Implement the method to extract all dependencies and write to a file under root

- [x]  Add unit tests

### CODE REVIEW

**Comment by fuhuifang** [comment]
This code assumes `conda_dependencies`, `conda_dependencies_file`, `pip_requirements_file` can co-exist in yaml. Please let me know if this is not the case.

**Review by haozhestat — CHANGES_REQUESTED** [review]

File: shrike/build/commands/prepare.py [hunk 1]
```diff
@@ -587,6 +589,77 @@ def is_in_additional_includes(
...
+
+    def _extract_python_package_dependencies(self, requirements):
+        pip_dependencies = []
+        if "dependencies" in requirements:
+            dependencies = requirements.get("dependencies")
+            for dependencies_item in dependencies:
+                if isinstance(dependencies_item, dict) and "pip" in dependencies_item:
+                    pip_dependencies = dependencies_item["pip"]
+                    for pip_dependencies_item in pip_dependencies:
+                        if pip_dependencies_item.startswith("--index-url"):
```
Comment (dkmiller) [inline shrike/build/commands/prepare.py:659]: You can actually keep this (it's ok if the generated requirements.txt file includes `--index-url` lines).

Those won't do anything "funny" with component governance.

File: shrike/build/commands/prepare.py [hunk 2]
```diff
@@ -587,6 +589,77 @@ def is_in_additional_includes(
...
+                        os.makedirs(cur_path)
+                        with open(
+                            os.path.join(cur_path, "requirements.txt"), "w"
+                        ) as file:
+                            for req in pip_dependencies:
+                                if req:
+                                    file.writelines(req + "\n")
+        return id
+
+    def _extract_python_package_dependencies(self, requirements):
```
Comment (dkmiller) [inline shrike/build/commands/prepare.py:651]: Is this extracting from requirements.txt or Conda files? The parameter name suggests former, but implementation suggests the latter.

File: shrike/build/commands/prepare.py [hunk 3]
```diff
@@ -587,6 +589,77 @@ def is_in_additional_includes(
...
+                        with open(
+                            os.path.join(
+                                component_repo, spec_conda["pip_requirements_file"]
+                            )
+                        ) as pip_requirements_file:
+                            pip_dependencies += pip_requirements_file.readlines()
+
+                    if pip_dependencies:
+                        pip_dependencies = [
+                            x.replace("\n", "") for x in pip_dependencies
```
Comment (dkmiller) [inline shrike/build/commands/prepare.py:635]: Are you sure this is needed? I think newlines in the generated requirements.txt file are ok.

File: shrike/build/commands/prepare.py [hunk 4]
```diff
@@ -587,6 +589,77 @@ def is_in_additional_includes(
...
+    def _create_requirements_file(self, component_files):
+        id = str(uuid.uuid4())
+        dependencies_path = os.path.join(
+            self.config.working_directory, "component_dependencies_" + id
+        )
+        log.info(f"Writing Python package dependencies to path {dependencies_path}")
+        os.makedirs(dependencies_path)
+        for component in component_files:
+            component_repo = Path(component).parent
+            with open(component, "r") as spec_file:
```
Comment (dkmiller) [inline shrike/build/commands/prepare.py:601]: Suggestion: have a "try / except" around all these `open` statements where you catch the `filenotfounderror` and register an error.

That way, wrong path to environment.yml doesn't catastrophically break the build.

**Review by dkmiller — CHANGES_REQUESTED** [review]

File: tests/tests_build/commands/test_prepare.py [hunk 5]
```diff
@@ -793,3 +793,239 @@ def test_add_repo_and_last_pr_to_tags():
...
+        lines = file.readlines()
+        assert lines == ["azureml-defaults\n", "scikit-learn==0.24.2\n"]
+
+
+@pytest.mark.order(-1)
+def test_create_requirements_file_for_single_component_no_dependencies(caplog):
+    clean()
+    prep = prepare.Prepare()
+    prep.config = Configuration()
+    prep.attach_workspace(TESTING_WORKSPACE)
```
Comment (dkmiller) [inline tests/tests_build/commands/test_prepare.py:1003]: A bit curious, why is workspace attach needed here? Ditto with some of the other tests?

Comment (fuhuifang) [reply tests/tests_build/commands/test_prepare.py]: Thanks for catching this. Just removed all necessary calls.

**Review by dkmiller — APPROVED** [review]

**Review by haozhestat — APPROVED** [review]
After this version is released, please upload the wheel file to the Polymer feed. Thanks, @fuhuifang.

File: shrike/__init__.py [hunk 6]
```diff
@@ -5,4 +5,4 @@
 Python utilities to aid "compliant experimentation" - training
 machine learning models without seeing the training data.
 """
-__version__ = "1.2.2"
+__version__ = "1.3.2"
```
Comment (haozhestat) [inline shrike/__init__.py:8]: @fuhuifang : I just realized that we skipped the "1.2.0", "1.3.0" and "1.3.1" versions of shrike. The current latest version of shrike is "1.3.2". No actions are required now. Let's follow the versioning convention in the future.

image

Comment (fuhuifang) [reply shrike/__init__.py:8]: Thanks Haozhe for the reminder! I didn't realize this at all. Will pay attention next time 😄

Comment (haozhestat) [reply shrike/__init__.py:8]: @fuhuifang : No problem at all. My bad. I should have caught it when I reviewed the PR.

**Pull request merged**